cleanup/stop accessioning rake tasks

peetucket opened the pull request:
## Why was this change made? 🤔

Fixes sul-dlss/common-accessioning#1060 - create a rake task that "resets"/cleanups the in process accession of a given druid.  Also add a rake task that allows you to pass in a CSV with multiple druids.

Todo:
- [ ] add documentation to DSA DevOpsDocs?  Elsewhere?

## How was this change tested? 🤨

A bit on localhost and stage.  Tested by Andrew on stage

jmartin-sul commented on lib/tasks/cleanup.rake:
```diff
@@ -0,0 +1,114 @@
+      specific_backup_path = File.join(base_backup_path, File.basename(path)) # e.g. /dor/staging/stopped/ab123bc4567/workspace
+      $stdout.puts "...found #{content_path.path}: copying to #{specific_backup_path} and then deleting"
+      next if dryrun
+
```
a message about "actually moving/deleting the stuff" might be nice if the dry run guard clause is passed

peetucket replied:
how about if I show a message before each druid is run (even when run in a loop) to make it clear when you are in dry run mode that nothing is actually being done

jmartin-sul reviewed (approved):
left some nitpicky suggestions/questions inline, and a misgiving about the check to see whether a druid is preserved, but i might've talked myself out of that.  still interested in whether you think there's anything to my concerns.

the overall approach to moving content and deleting workflows seems good to me.  i guess my one larger misgiving is that the `cleanup_druid` method is of course somewhat complex and can delete some important files and DB rows, so it seems like it'd be nice to put it in app and give it some test coverage.  as a bonus, that'd make it usable in scripting from rails console and not just the shell.  OTOH, andrew tested it, and if things about it fall out of date, it doesn't seem like it'll fail in terrible ways.  its file deletes are constrained and it goes through the workflow client to delete workflows.

given all of that, i think i'd lean toward merging this as-is, and if any of my concerns/suggestions have merit, dealing with them in follow-on issues/PRs.

peetucket commented on lib/tasks/cleanup.rake:
```diff
@@ -0,0 +1,114 @@
+  # rubocop:disable Metrics/AbcSize
+  # rubocop:disable Metrics/MethodLength
+  def cleanup_druid(druid, dryrun: false)
+    backup_path = '/dor/staging/stopped' # path to move content before deleting from cleanup_paths
```
Like a parameter you can pass into the rake task?

peetucket replied:
Oh yes, that makes sense.  I'll do that.

peetucket commented on lib/tasks/cleanup.rake:
```diff
@@ -0,0 +1,114 @@
+    object = CocinaObjectStore.find(druid)
+
+    # Verify this is a item type object
+    raise 'object is not an item: cannot proceed' unless object.dro?
```
That was the thought process, but I presume there is no reason this cleanup couldn't run on collections -- so I'll remove this check

peetucket commented on lib/tasks/cleanup.rake:
```diff
@@ -0,0 +1,114 @@
+
+    # If `preservationIngestWF#complete-ingest` == waiting, then a step in this workflow is likely in error
+    #  and we should stop, since extra remediation may be needed
+    raise "v#{object.version} of the object has preservationIngestWF#complete-ingest set to waiting: cannot proceed" if WorkflowClientFactory.build.workflow_status(druid:, workflow: 'preservationIngestWF', process: 'complete-ingest') == 'waiting'
```
Effectively we have two checks suggested by Andrew:

The first is checking the `openable` state, which using the `VersionService` does various checks which effectively boil down to "is this in preservation?".  

So if something is not openable, it is:

- not yet in preservation (because it hasn't been accessioned at all yet)
- it is in the middle of accessioning a new version 

These are the two situations in which a cleanup is allowed.  So basically, if we can open it, then we must be in a state that is not allowed for cleanup.

The second is trying to catch the edge case you describe above, it got stuck very late in the preservation process such that SDR thinks it is openable, but there is really a problem in `preservationIngestWF'.  This requires some closer examination before cleanup should occur, and we stop.

So basically we do both a version check and a workflow check.

I did alter the workflow check a bit so that instead of looking for the last step in `preservationIngestWF` being `waiting`, I instead look for it NOT being `completed` ... this is so that the lasts step being skipped or error will also block the cleanup, as unlikely as either of those would be.

peetucket commented on lib/tasks/cleanup.rake:
```diff
@@ -0,0 +1,114 @@
+
+    rows.each do |row|
+      druid = row.first
+      $stdout.puts druid
```
Added some whitespace but rubocop DENIED some of it

peetucket replied:
Got it, will add.

peetucket commented:
> the overall approach to moving content and deleting workflows seems good to me. i guess my one larger misgiving is that the `cleanup_druid` method is of course somewhat complex and can delete some important files and DB rows, so it seems like it'd be nice to put it in app and give it some test coverage. as a bonus, that'd make it usable in scripting from rails console and not just the shell. OTOH, andrew tested it, and if things about it fall out of date, it doesn't seem like it'll fail in terrible ways. its file deletes are constrained and it goes through the workflow client to delete workflows.

The copying/backup logic is probably the most novel, the other actions are using existing services (e.g. for deleting folders or workflows)...but I can see how putting this in a class that could be testable is wise.  I had some verbose output to stdout so you can see what's happening when the task, how do you feel about that style of output in a service class?

peetucket commented:
OK, now I moved all of the logic to the `CleanupService`, added tests, and then shrunk the rake task down to just wrapping the call to CleanupService.

jmartin-sul reviewed (approved):
✅ again, thanks for the refactor and tests.  leaving to you to merge in case you or andrew still need/want to test the refactored code on stage or QA

peetucket commented:
> ✅ again, thanks for the refactor and tests. leaving to you to merge in case you or andrew still need/want to test the refactored code on stage or QA (would not hold up merging for either of my suggestions, would probably rather seem them as follow-ons if you think they're worth the effort, and fine to skip them)

Thanks ... will merge as is, then add documentation, then do a small follow on PR for your good suggestions.
